main/src/ca/uwaterloo/flix/language/phase/unification/FastBoolUnification.scala

```
@@ -0,0 +1,1715 @@
+
+    // INVARIANT: The current substitution has been applied to BOTH the unsolved AND pending equations.
+    while (pending.nonEmpty) {
+      // Extract the next equation and update the pending equations.
+      val e = pending.head
+      pending = pending.tail
+
+      e match {
+        case Equation(Term.Var(x), rhs, _) if !rhs.freeVars.contains(x) =>
+          // Case 1: We have found an equation: `x ~ t` where `x` is not free in `t`.
+          // Construct a singleton substitution `[x -> y]`.
+          val singleton = BoolSubstitution.singleton(x, rhs)
+
+          // Apply the singleton substitution to the unsolved equations.
+          pending = singleton(pending)
+
+          // Apply the singleton substitution to the unsolved equations.
+          unsolved = singleton(unsolved)
+
+          // Compose the new substitution with the current substitution.
+          subst = singleton @@ subst
+
+        case _ => unsolved = e :: unsolved
+      }
+    }
+
+    // Reverse the unsolved equations to ensure they are returned in the original order.
+    (unsolved.reverse, subst)
+  }
+
+  /**
+   * Given a unification equation system `l` computes a most-general unifier for all its equations.
+   *
+   * If multiple equations are involved then we try to solve them in different order to find a small substitution.
+   */
+  private def boolUnifyAllPickSmallest(l: List[Equation]): BoolSubstitution = {
+    // Case 1: We have at most one equation to solve: just solve immediately.
+    if (l.length <= 1) {
+      return boolUnifyAll(l)
+    }
+
+    // Case 2: We have too many equations. We cannot hope to compute with all their permutations: just solve immediately.
+    if (l.length > PermutationLimit) {
+      return boolUnifyAll(l)
+    }
+
+    // Case 3: We have a reasonable number of permutations. We solve all of them and pick the smallest substitution.
+    val results = l.permutations.toList.map {
+      case p => (p, boolUnifyAll(p))
```

You can give the current min size to boolUnifyAll so it can fail faster when the size becomes too big. Feels like that could give a lot of performance

---

I think lots of tricks can be played, but I need more than one example to evaluate them :)